File: apps/mobile/src/features/core/native-core-adapters/message.ts
import type { Message } from '@glimpse/shared';

import { nativeCoreBridgeHelpers } from '../native-core-bridge';
import type { BridgeCoreClient } from '../types';

import type { NativeAdapterDeps } from './common';

const { fromNitroMessage, toMessagePatch, toNitroMessage } = nativeCoreBridgeHelpers;

export function createMessageAdapter(
  deps: NativeAdapterDeps,
): Pick<
  BridgeCoreClient,
  'listConversationMessages' | 'addMessage' | 'updateMessage' | 'deleteMessage'
> {
  const { fallbackClient, runCoreAsync } = deps;

  return {
    async listConversationMessages(conversationId: string): Promise<Message[]> {
      return runCoreAsync(
        async (client) =>
          (await client.listConversationMessages(conversationId)).map(fromNitroMessage),
        () => fallbackClient.listConversationMessages(conversationId),
      );
    },

    async addMessage(message: Message): Promise<Message> {
      return runCoreAsync(
        async (client) => fromNitroMessage(await client.addMessage(toNitroMessage(message))),
        () => fallbackClient.addMessage(message),
      );
    },

    async updateMessage(
      messageId: string,
      patch: Partial<Message>,
    ): Promise<Message> {
      return runCoreAsync(
        async (client) =>
          fromNitroMessage(await client.updateMessage(messageId, toMessagePatch(patch))),
        () => fallbackClient.updateMessage(messageId, patch),
      );
    },

    async deleteMessage(messageId: string, deletedAt: number): Promise<void> {
      await runCoreAsync(
        (client) => client.deleteMessage(messageId, deletedAt),
        () => fallbackClient.deleteMessage(messageId, deletedAt),
      );
    },
  };
}

File: packages/core-rust/src/application/message.rs
use crate::error::Result;
use crate::models::{Message, MessagePatch};

use super::SharedCore;

impl SharedCore {
    pub fn list_conversation_messages(&self, conversation_id: &str) -> Result<Vec<Message>> {
        self.client().list_conversation_messages(conversation_id)
    }

    pub fn add_message(&self, message: &Message) -> Result<Message> {
        self.client().add_message(message)
    }

    pub fn update_message(&self, message_id: &str, patch: &MessagePatch) -> Result<Message> {
        self.client().update_message(message_id, patch)
    }

    pub fn delete_message(&self, message_id: &str, deleted_at: i64) -> Result<()> {
        self.client().delete_message(message_id, deleted_at)
    }
}
